qldpc/codes.py

```diff
@@ -421,6 +507,61 @@ def num_logical_qubits(self) -> int:
         """Number of logical qubits encoded by this code."""
         return self.code_x.dimension + self.code_z.dimension - self.num_qubits
 
+    def _assert_qubit_code(self) -> None:
+        if self._field_order != 2:
+            raise ValueError("Attempted to call a qubit-only method with a non-qubit code.")
+
+    """
+    Code tranformation methods — Conjugation and Shifting.
+
+    (i) Conjugation — A CSSCode can specify which data qubits should be hadamard-transformed
+    before/after syndrome extraction, thereby transforming the operators
+    that address a specified data qubit as (X,Y,Z) <--> (Z,Y,X).
+
+    (ii) Shifting — A CSSCode can also "shift" the Pauli operators on a qubit, moving vertically
+    along the following table:
+
+    ―――――――――――
+    | XY | YX |
+    ―――――――――――
+    | XZ | ZX |
+    ―――――――――――
+    | YZ | ZY |
+    ―――――――――――
+
+    Qubit shifts are specified by a dictionary mapping qubit_index --> shift_index, where
+    shift_index = +1, 0, and -1 mod 3 respectively refer to the top, middle, and bottom rows of the
+    table.
+
+    Physically, a shift of +1 and -1 respectively correspond to conjugating the Pauli operators
+    addressing a qubit by sqrt(X) and sqrt(Z) rotations.
+    """
+
+    # Given a parity check matrix, swap X and Z operators on the qubits to conjugate.
+    # TODO: Does this work for general stabilizer code (over F_2)? If so, move to QuditCode class?
+    @classmethod
+    def conjugate(cls, matrix: IntegerMatrix | None, conj: slice | Sequence[int]) -> IntegerMatrix:
+        matrix[:, :, conj] = np.roll(matrix[:, :, conj], 1, axis=1)
+        return matrix
+
+    @classmethod
+    def shift(cls, matrix: IntegerMatrix | None, shifts: dict[int, int]) -> IntegerMatrix:
```

Comment:
1. Let's merge these two methods into something like a `deform` method that transforms a parity check matrix by applying local Clifford deformations.
2. This is definitely general and can live in `QuditCode`.